src/simulations/collect.jl

```diff
@@ -164,6 +165,128 @@ function run!(
 end
 
+"""
+    offline_run!(model, agent_step! [, model_step!], n::Integer; kwargs...)
+    offline_run!(model, agent_step!, model_step!, n::Function; kwargs...)
+
+Do the same as [`run`](@ref), but write output to a CSV file instead of collecting
+it into an in-memory dataframe.
+Useful when the amount of collected data is expected to exceed the memory available
+during execution.
+
+## Keywords
+* `adata_savefile="adata.csv"` : a file to write agent data on.
+* `mdata_savefile="mdata.csv"`: a file to write the model data on.
+* `writing_interval=1` : write to file every `writing_interval` times data collection \
+is triggered. If the `when` keyword is not set, this corresponds to writing to file \
+every `writing_interval` steps; otherwise, the data will be written every \
+`writing_interval` times the `when` condition is satisfied \
+(the same applies to `when_model`).
+"""
+function offline_run! end
+
+offline_run!(model::ABM, agent_step!, n::Int = 1; kwargs...) =
+    offline_run!(model::ABM, agent_step!, dummystep, n; kwargs...)
+
+function offline_run!(
+    model,
+    agent_step!,
+    model_step!,
+    n;
+    when = true,
+    when_model = when,
+    mdata = nothing,
+    adata = nothing,
+    obtainer = identity,
+    agents_first = true,
+    showprogress = false,
+    writing_interval = 1,
+    adata_savefile = "adata.csv",
+    mdata_savefile = "mdata.csv"
+)
+
+    df_agent = init_agent_dataframe(model, adata)
+    df_model = init_model_dataframe(model, mdata)
+    if n isa Integer
+        if when == true
+            for c in eachcol(df_agent)
+                sizehint!(c, n)
+            end
+        end
+        if when_model == true
+            for c in eachcol(df_model)
+                sizehint!(c, n)
+            end
+        end
+    end
+
+    close(open(adata_savefile, "w"))
+    close(open(mdata_savefile, "w"))
+    model_appendtocsv = false
+    agent_appendtocsv = false
+
+    agent_count_collections = 0
+    model_count_collections = 0
+    agent_collected = false
+    model_collected = false
+
+    s = 0
+    p = if typeof(n) <: Int
+        ProgressMeter.Progress(n; enabled=showprogress, desc="run! progress: ")
+    else
+        ProgressMeter.ProgressUnknown(desc="run! steps done: ", enabled=showprogress)
+    end
+    while until(s, n, model)
+        if should_we_collect(s, model, when)
+            collect_agent_data!(df_agent, model, adata, s; obtainer)
+            agent_count_collections += 1
+            agent_collected = true
+        end
+        if should_we_collect(s, model, when_model)
+            collect_model_data!(df_model, model, mdata, s; obtainer)
+            model_count_collections += 1
+            model_collected = true
+        end
+        step!(model, agent_step!, model_step!, 1, agents_first)
+        s += 1
+
+        if model_collected && model_count_collections % writing_interval == 0
+            AgentsIO.CSV.write(mdata_savefile, df_model, append=model_appendtocsv)
```

Comment:
Please replace all `,` with `;` when going from positional arguments to keyword arguments.